### ISSUE TITLE
sql: enhance the namespace caching story in light of multiple schemas

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#44802]
tldr: we've started to introduce other schemas than `public` and this triggers an unexpected extra KV lookup in a common case, hurting performance. We need to find a solution.

Table of contents for the below:
- Problem description
- Observation: the problem is the crdb 1.1 compat mode
- Alt 1: revert changes in 20.1
- Alt 2: deprecate the crdb 1.1 compat mode
- Alt 3: 2 classes of behavior (@knz)
- Alt 4: negative resolution cache (@petermattis)

### Problem description

1. familiarize yourself the pg name resolution rules, see [this intro](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20180219_pg_virtual_namespacing.md#reference-level-explanation) in RFC then [this section](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20180219_pg_virtual_namespacing.md#changes-to-algorithm-a1) of the RFC.
2. the problem is the case of `X.Y`. Postgres says this is supposed to mean "schema X, table Y". In crdb, we have a fallback "if schema X is not found, try to find a db named X" for back-compat with crdb 1.1 and prior.
3. that special case 2 was _previously_ optimized by saying:

   a. if the word "X" is exactly "`public`", do the regular lookup which will hit the namespace cache.
   b. if the word "X" is exactly one of the virtual schemas, do the regular lookup (which doesn't need KV)
   c. *otherwise* assume "X" is a database, and perform lookup "X.public.Y" which also will hit the namespace cache in the common case.

4. with arbitrary other schema names than public or the fixed list of vschemas, we must extend the rule as follows:

   a. same as 3a above
   b. same as 3b above
   c. **look up a schema in KV** named "X" in system.namespace
   d. **only if a schema is not found** assume "X" is a database, lookup "X.public.Y" and hit the cache in the common case.

5. the problem is that in a relatively common case where people use "db.table", the word "db" is not a schema, so step 4c above **incurs an extra KV lookup that's bound to fail, in the common case.**

Tada we've pessimized the common case.

We're going to assume that the extra lookup in the common case is unacceptable for the rest of this issue.

### Observation: the problem is the back-compat

If we did not accept "db.table" as alias for "db.public.table" then  applications would write "public.table" or "myschema.table" and seeing the word "public" we'd hit the namespace cache in the common case.

So really we shot ourselves in the foot by designing a weird non-pg behavior.

Note/observation: pg does not "need" the db.table notation because it's not possible with pg to establish a SQL conn without a "current db". If we did not allow db-less conns in crdb from the start, we wouldn't have designed it this way.

### Alt 1: walk back on the 20.1. change

Cancel the addition of additional schemas entirely and go back to 19.2 behavior.

Pros: simple to do
Cons: blocks many important new features (temp tables, user-defined schemas)

### Alt 2: outlaw the notation "db.table"

This needs a deprecation cycle because we can't just drop everything without some prior notice.

1. make the new schema features opt-in in 20.1 behind a cluster setting / session flag
2. when not opted in, the extra lookup is not performed so things are fast
3. when opted in, the extra lookup is performed and things are slow(er) 
4. announce the syntax "db.table" is now deprecated in 20.1.  Announce that we're going to require a "current db" in all cons starting in 20.2
5. make the thing opt out in 20.2 and require a current db
6. remove the  "db.table" notation in 21.1

Pros: our usual deprecation process, eventually clearer story compared to pg

Cons: the notation "db.table" is deeply anchored with our testing code, 3rd party apps etc. Unclear how painful the transition will be

Cons (bigger): we can't really publicize the new feature (temp tables, user schemas) in 20.1 and 20.2 because they hurt performance in a noticeable way when enabled.

### Alt 3: acknowledge two classes of client conns

Let's make use of an observation "from the field": *clients that use the 'db.table' syntax are mostly those that use a db-less connection*.

Conversely, clients that use a db-full connection tend to not use it.

That is, **the "common" case** is split in two groups with little overlap:

- sessions with empty `database`/`current_catalog`, using `db.table` or `vschema.table`
- session with non-empty `database`, using `table` or `vschema.table` or `altschema.table`

Hence a proposal for an alternate approach:

- make name resolution when `database` is non-empty exactly like pg, and avoid the nasty extra lookup.
- make name resolution when `database` is empty using a different algorithm using the following order for X.Y:
  1. try X as virtual schema X (not real schema) first
  2. only then try database X with schema `public`
  3. if not found, fail with error

The keen eye recognizes that the 2nd algorithm is unable to use non-public, non-virtual schemas. I think that's acceptable: it's for legacy code using db-less conns which is not expecting those extra schemas anyway.

### Alt 4: negative resolution cache

Came from @petermattis in discussion.

Keep the algorithm 4 a-b-c-d above, but introduce a cache of failed KV lookups on step 4.c.

That cache would say "in X.Y, X is not a schema name".

That cache would be keyed by ..... (TBD)

That cache would be invalidated by .... (TBD)

Pros: keeps the behavior consistent across all connection mode, in contrast to alt 3. Remains backward-compat in all cases.

Cons: introduces a new async mechanism.

Cons: that new cache's invalidation policy will be another invalidation dance, for which we've not been very good in the past.

Jira issue: CRDB-6330

**Comment by knz** [comment]
cc @ajwerner @otan @solongordon

**Comment by petermattis** [comment]
> a. same as 3a above
b. same as 3b above
c. look up a schema in KV named "X" in system.namespace
d. only if a schema is not found assume "X" is a database, lookup "X.public.Y" and hit the cache in the common case.

The extra KV lookup is in step c, right? And it continues to occur because we do not have a cache saying "X" is not a schema? I'm not seeing why this requires a KV lookup on each operation. Can we add such a cache (and would a cache here help)? I've forgotten the details of how the namespace cache works, so pardon the naive question.

> Alt 1: walk back on the 20.1. change

Can this be done while another solution is worked out? I imagine the performance hit is unacceptable to release with, and it will get in the way of other team's doing performance testing on master.

> Alt 2: outlaw the notation "db.table"
>
> This needs a deprecation cycle because we can't just drop everything without some prior notice.

My 2 cents: this will be nearly infeasible to do.

**Comment by knz** [comment]
> And it continues to occur because we do not have a cache saying "X" is not a schema? I'm not seeing why this requires a KV lookup on each operation.

So that idea is to create a "negative answer" cache  (which we don't have today), for which we'll also need an invalidation policy (which I don't know how to define yet).

I'd create an "Alternative 4" section for that. Can you flesh the semantics of that cache out, especially the invalidation?

**Comment by petermattis** [comment]
> So that idea is to create a "negative answer" cache (which we don't have today), for which we'll also need an invalidation policy (which I don't know how to define yet).

Yes, I was suggestion a negative cache. 

> I'd create an "Alternative 4" section for that. Can you flesh the semantics of that cache out, especially the invalidation?

Yeah, invalidation is the tricky part. Invalidation is always the tricky part. Can you point me to the current namespace lookup code? Everything in `pkg/sql` has moved around from the last time I worked in there.

**Comment by knz** [comment]
> Can you point me to the current namespace lookup code?

The name resolution algorithm is generic and takes a `sql.SchemaAccessor` as accessor for the namespaces and descriptors.

We implement SchemaAccessor by layering a `LogicalSchemaAccessor` (pkg/sql/logical_schema_accessors.go`, recognizes virtual schemas) over a `CachedPhysicalAccessor` (pkg/sql/physical_schema_accessors.go`, implements caching) over a `UncachedPhysicalAccessor` (same file).

Long story short the cache is implemented by the `TableCollection` object (`table.go`). This contains two separate caches currently, one for database lookups (poor/terrible invalidation generally) and one for table lookups (ok-ish invalidation).

**Comment by ajwerner** [comment]
It is very possible that I’m not yet awake but it seems like this discussion is missing the workaround we arrived at yesterday. In #44783, @otan found a nice compromise that allows us to continue to work with temp table schemas but not all schemas. Only schemas with the `pg_temp` prefix will result in the lookup. 

We still have a thorny problem to solve but it seems like it can be deferred to 20.2 and a more general implementation of schemas.

**Comment by petermattis** [comment]
From my reading, it looks like `databaseCache` is reset whenever the `SystemConfig` is gossiped:

```
func (dc *databaseCacheHolder) updateSystemConfig(cfg *config.SystemConfig) {
	dc.mu.Lock()
	dc.mu.c = newDatabaseCache(cfg)
	dc.mu.cv.Broadcast()
	dc.mu.Unlock()
}
```

The `SystemConfig` will be gossiped whenever there is a change to the `system.namespace` table. I don't see why we can't cache negative lookups in the database cache. And it seems like an exactly analogous cache could be used for schema lookups. Completely invalidating these caches whenever `SystemConfig` is gossiped is itself a bit of a wart (@ajwerner desires something better in the future), yet that is our current state. Am I missing anything that would make caching negative entries for both databases and schemas problematic? One change from the current behavior, is that right now a newly created database (or schema) is visible immediately on other nodes in advance of gossip propagating. If we cached negative entries that wouldn't be the case. And this would happen in practice for the schema cache as we'd frequently be caching the lookups for database names which are not actually schema names. My guess is this won't be problematic, but others should weigh in with their opinions.

**Comment by knz** [comment]
@ajwerner thanks for making me aware of #44783 I had missed that.
This means that the severity can be decreased for now.

We still need to address the issue, because user-defined schemas will eventually become a thing.

**Comment by otan** [comment]
adding a note here, having this turned off also hurts ONE table name queries of form just `X` if pg_temp is enabled, e.g. just `select * from <table>`.

this is because it tries to include all schemas in the search path (such as `pg_temp.table`) before it tries `public.table`.

see benchmark results: https://github.com/cockroachdb/cockroach/pull/44752#issuecomment-583133633

**Comment by rafiss** [comment]
We should probably revisit this, since user-defined schemas did become a thing. Note that  https://github.com/cockroachdb/cockroach/pull/44783 was undone in https://github.com/cockroachdb/cockroach/pull/50664. We've seen quite a bit of fallout in support cases (e.g. see [this postmortem](https://cockroachlabs.atlassian.net/wiki/spaces/SQLEXP/pages/1328906263/CloudKitchens+authentication+performance+post-mortem), [another postmortem](https://cockroachlabs.atlassian.net/wiki/spaces/MC/pages/1196982501/2020-12-17%3A+AllSaints+Connection+Issues+and+Escalation), https://github.com/cockroachlabs/support/issues/791, https://github.com/cockroachlabs/support/issues/775)

The optimization in https://github.com/cockroachdb/cockroach/pull/58674 was added as a result of the above -- don't look up the `""` database name. This name would be looked up if you were not connected to any DB and you used the `db.table` syntax in a query. This addresses note 2 in @otan's comment in some cases (most importantly in the internal executor). However, the roundtrip benchmarks still show that querying `system.<x>` incurs an extra lookup if a `current DB` is set. https://github.com/cockroachdb/cockroach/blob/a5bea8c7fbb979de7e0014bc7ced35a34a209dfc/pkg/bench/ddl_analysis/testdata/benchmark_expectations#L66

I think going ahead with implementing the negative cache for both schema names and database names would be quite useful.

**Comment by rafiss** [comment]
Seems like schema name lookup does have short-circuiting logic, here: https://github.com/cockroachdb/cockroach/pull/53290/files#diff-72793be0a520947d817964df62478b8aca5aeedae30f3e1f6f6bb5eccaf0f360R498

I don't think there's a negative cache for database names, but that doesn't seem as important

**Comment by rafiss** [comment]
I consider this done based on my last update. I don't think further changes are worth the complexity or effort. Let me know if you disagree.
